### DISCUSSION TITLE
helping MUMPS configure find LAPACK

### DISCUSSION THREAD

**Discussion by rdzman** [coin-or/Ipopt#544] [Q&A] [answered]
I'm trying to add an IPOPT MEX interface to a [GNU Octave Docker image](https://github.com/gnu-octave/docker). Using `apt-get coinor-libipopt-dev` resulted in a MEX file that causes a segfault. I'm guessing it may be because of mismatched BLAS, LAPACK, etc. (though this is pure conjecture ... not sure how to confirm).

So, I'm now trying to build IPOPT from scratch using `coinbrew` and running into an issue with the `configure` for MUMPS not being able to find LAPACK. I'm assuming it's already been used to build Octave and is around somewhere, but I'm not sure where. According to [this docker file](https://github.com/gnu-octave/docker/blob/main/build-octave-6.docker), they are installing OpenBLAS and SuiteSparse in `/usr`. And `/usr/lib` does contain `libopenblas.so`, for example, but I'm not familiar with which library specifically would contain LAPACK.

I've tried setting the `with_lapack_lflags` env variable to several different values with no luck. I always get an error like, e.g. ...
```
configure: error: Cannot link to user-specified Lapack  -L/usr/lib -llibopenblas.so.
```

Any suggestions for what `with_lapack_lflags` should be?

**Comment by rdzman** [comment, marked as answer]
Ok, I figured that out. Setting `with_lapack_lflags` to `-lopenblas` allows it to configure and build everything just fine now. However, it's segfaulting on the unit tests now.

Any suggestions on how to track that down?


```
##################################################
### Building Ipopt 3.14.4 
##################################################


##################################################
### Running Ipopt unit test 
##################################################

./run_unitTests: line 21: 18957 Segmentation fault      $@ > tmpfile 2>&1
./run_unitTests: line 21: 18972 Segmentation fault      $@ > tmpfile 2>&1
./run_unitTests: line 21: 18987 Segmentation fault      $@ > tmpfile 2>&1
./run_unitTests: line 21: 19002 Segmentation fault      $@ > tmpfile 2>&1
./run_unitTests: line 21: 19017 Aborted                 $@ > tmpfile 2>&1
./run_unitTests: line 21: 19035 Segmentation fault      $@ > tmpfile 2>&1
./run_unitTests: line 21: 19050 Segmentation fault      $@ > tmpfile 2>&1
./run_unitTests: line 21: 19065 Segmentation fault      $@ > tmpfile 2>&1
./run_unitTests: line 21: 19080 Segmentation fault      $@ > tmpfile 2>&1
make[1]: *** [Makefile:816: test] Error 1
make: *** [Makefile:848: test] Error 2
```

**Reply by rdzman** [reply]
A bit more info ... it looks like the verbose output of the Java based test has the most info ...
```
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00007fa5f7089d87, pid=22130, tid=22131
#
# JRE version: OpenJDK Runtime Environment (11.0.11+9) (build 11.0.11+9-Ubuntu-0ubuntu2.20.04)
# Java VM: OpenJDK 64-Bit Server VM (11.0.11+9-Ubuntu-0ubuntu2.20.04, mixed mode, sharing, tiered, compressed oops, g1 gc, linux-amd64)
# Problematic frame:
# C  [libopenblas.so.0+0x179dd87]  dcopy_k_HASWELL+0x387
```

And another test says ...
```
Program received signal SIGSEGV: Segmentation fault - invalid memory reference.

Backtrace for this error:
#0  0x7f04294c2d5a
#1  0x7f04294c1ef5
#2  0x7f04292f620f
#3  0x7f04280cbd87
```

**Reply by svigerske** [reply]
It looks like they build OpenBlas with an interface for 64-bit integers for Octave:

https://github.com/gnu-octave/docker/blob/dd3d3b6f4b9202423275424ecb77c7b3bb57f7ed/build-octave-6.docker#L108-L122

But Ipopt and MUMPS assume 32-bit integers (or actually `int`) by default.
That's the same problem users sometimes have with MATLAB/Ipopt interfaces.

If you build MUMPS and Ipopt against a Blas/Lapack with `int` for integers instead, you should use static libraries for Blas/Lapack to avoid that the Blas/Lapack routines from Octave are still used (when they are loaded first).   
Or you could build Ipopt and MUMPS to use 64-bit integers, too (https://coin-or.github.io/Ipopt/INSTALL.html#INT64_BUILD). But this comes with increased memory usage as _every_ integer in Ipopt will be 64-bit then.

**Reply by siko1056** [reply]
Here the author of the Octave Docker images :hand:  I just finished answering the question of @rdzman :

https://github.com/gnu-octave/docker/issues/5#issuecomment-994291575

@svigerske your analysis of the problem is correct and the solution is provided above.  You have a great Ipopt documentation regarding ilp64 compilation :+1:  Sorry for the duplicated effort looking into this (@rdzman please [avoid crossposing](https://en.wikipedia.org/wiki/Crossposting), in FLOSS one cannot always receive a reply within "one unpaid business day" :innocent: ).

**Reply by rdzman** [reply]
Thanks a ton to both of you, and sorry for the duplicate effort. Now I can build a working IPOPT, but the MEX build is failing. I'll post all followups in gnu-octave/docker#5.
